File: stocks/src/core/infrastructure/repository.ts
import * as dynamo from "@aws-sdk/client-dynamodb";
import * as dynamotool from "@aws-sdk/util-dynamodb";
import * as Pino from "pino";

import * as repomdl from "./../model/repository";
import * as conf from "./../model/conf";
import { Stock } from "./../model/stock";
import { StockNotFoundError } from "./../model/errors";

const logger = Pino.default({
  level: "debug",
});

const dateFrom = (dateObj: any): Date => {
  return new Date(dateObj);
};

const createSubmitTxInput = (
  // TODO: next - the exception is happening here
  changeSet: repomdl.ChangeSet
): dynamo.TransactWriteItemsInput => {
  const cnf = conf.get();

  const put1: dynamo.Put = {} as dynamo.Put;
  put1.TableName = cnf.tableIncomingEvents;
  put1.Item = dynamotool.marshall(changeSet.incoming);
  put1.ConditionExpression = "attribute_not_exists(id)";

  const txitem1: dynamo.TransactWriteItem = {} as dynamo.TransactWriteItem;
  txitem1.Put = put1;

  const put2: dynamo.Put = {} as dynamo.Put;
  put2.TableName = cnf.tableStocks;
  const updatedStock = {
    ...changeSet.updatedStock,
    date: dateFrom(changeSet.updatedStock.date).getTime(),
  };
  put2.Item = dynamotool.marshall(updatedStock);

  if (changeSet.oldStockRev) {
    put2.ConditionExpression = "#rev == :rev_val";
    put2.ExpressionAttributeNames = {
      "#rev": "rev",
    };
    put2.ExpressionAttributeValues = dynamotool.marshall({
      ":rev_val": changeSet.oldStockRev,
    });
  }

  const txitem2: dynamo.TransactWriteItem = {} as dynamo.TransactWriteItem;
  txitem2.Put = put2;

  const put3: dynamo.Put = {} as dynamo.Put;
  put3.TableName = cnf.tableOutgoingEvents;
  const outgoing: any = changeSet.outgoing;
  outgoing.stock.date = dateFrom(changeSet.outgoing.stock.date).getTime();
  put3.Item = dynamotool.marshall(outgoing);

  const txitem3: dynamo.TransactWriteItem = {} as dynamo.TransactWriteItem;
  txitem3.Put = put3;

  const input: dynamo.TransactWriteItemsInput =
    {} as dynamo.TransactWriteItemsInput;
  input.TransactItems = [txitem1, txitem2, txitem3];

  return input;
};

export class Repository implements repomdl.Repository {
  protected client: dynamo.DynamoDB;

  constructor(client: dynamo.DynamoDB) {
    this.client = client;
  }

  async priceUpdatedEventExists(eventId: string): Promise<boolean> {
    const cnf = conf.get();

    const input: dynamo.GetItemCommandInput = {
      TableName: cnf.tableIncomingEvents,
      Key: {
        id: { S: eventId },
      },
    } as dynamo.GetItemCommandInput;

    const output = await this.client.getItem(input);
    if (!output.Item) {
      return false;
    }

    return true;
  }

  async loadStock(stockId: string): Promise<Stock> {
    const cnf = conf.get();

    const input: dynamo.GetItemCommandInput = {
      TableName: cnf.tableStocks,
      Key: {
        id: { S: stockId },
      },
    } as dynamo.GetItemCommandInput;

    const output = await this.client.getItem(input);
    if (!output.Item) {
      throw new StockNotFoundError();
    }

    return dynamotool.unmarshall(output.Item) as Stock;
  }

  async submit(changeSet: repomdl.ChangeSet): Promise<void> {
    const input: dynamo.TransactWriteItemsInput =
      createSubmitTxInput(changeSet);

    const output = await this.client.transactWriteItems(input);
    logger.info(output, "tx result");
  }
}

File: core/infrastructure/repository.go
package infrastructure

import (
	"context"
	"encoding/json"
	"os"

	"github.com/dc0d/reliable-messaging/core/model"

	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/feature/dynamodb/attributevalue"
	"github.com/aws/aws-sdk-go-v2/feature/dynamodb/expression"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
	"github.com/dc0d/wrapperr"
	"go.uber.org/zap"
)

type Repository struct {
	ctx    context.Context
	client *dynamodb.Client
}

func InitRepository(ctx context.Context) (*Repository, error) {
	result := &Repository{ctx: ctx}

	cfg, err := config.LoadDefaultConfig(result.ctx)
	if err != nil {
		return nil, wrapperr.WithStack(err)
	}

	cfg.Region = os.Getenv("OPS_AWS_REGION")

	result.client = dynamodb.NewFromConfig(cfg)

	return result, nil
}

func (repo *Repository) PriceUpdatedEventExists(eventID string) (bool, error) {
	ctx := repo.ctx
	client := repo.client

	table1 := tableIncomingEvents
	txgetitem1 := types.TransactGetItem{
		Get: &types.Get{
			TableName: &table1,
			Key: map[string]types.AttributeValue{
				"ID": &types.AttributeValueMemberS{
					Value: eventID,
				},
			},
		},
	}

	var input dynamodb.TransactGetItemsInput
	input.TransactItems = append(input.TransactItems, txgetitem1)

	output, err := client.TransactGetItems(ctx, &input)
	if err != nil {
		return false, wrapperr.WithStack(err)
	}

	var (
		prevIncoming model.PriceUpdated
	)

	if len(output.Responses) > 0 && output.Responses[0].Item != nil {
		if err := attributevalue.UnmarshalMap(output.Responses[0].Item, &prevIncoming); err != nil {
			return false, wrapperr.WithStack(err)
		}
	}

	if prevIncoming.ID == eventID {
		return true, nil
	}

	return false, nil
}

func (repo *Repository) LoadStock(stockID string) (result model.Stock, errResult error) {
	ctx := repo.ctx
	client := repo.client
	var (
		prevStock model.Stock
	)

	table := tableStocks
	txgetitem := types.TransactGetItem{
		Get: &types.Get{
			TableName: &table,
			Key: map[string]types.AttributeValue{
				"ID": &types.AttributeValueMemberS{
					Value: stockID,
				},
			},
		},
	}

	input := dynamodb.TransactGetItemsInput{}
	input.TransactItems = append(input.TransactItems, txgetitem)

	output, err := client.TransactGetItems(ctx, &input)
	if err != nil {
		errResult = wrapperr.WithStack(err)
		return
	}

	if len(output.Responses) > 0 && output.Responses[0].Item != nil {
		if err := attributevalue.UnmarshalMap(output.Responses[0].Item, &prevStock); err != nil {
			errResult = wrapperr.WithStack(err)
			return
		}

		result = prevStock
		return
	}

	errResult = model.ErrNotFound
	return
}

func (repo *Repository) Submit(changeSet model.ChangeSet) error {
	ctx := repo.ctx
	client := repo.client

	i1, err := txitem(changeSet.Incoming, tableIncomingEvents)
	if err != nil {
		return wrapperr.WithStack(err)
	}

	var i2 *types.TransactWriteItem
	if changeSet.OldStockRev != "" {
		op := expression.Name("Rev").Equal(expression.Value(changeSet.OldStockRev))
		expr, err := expression.NewBuilder().WithCondition(op).Build()

		if err != nil {
			return wrapperr.WithStack(err)
		}

		i2, err = txitem2(changeSet.UpdatedStock, tableStocks, expr)
		if err != nil {
			return wrapperr.WithStack(err)
		}
	} else {
		i2, err = txitem(changeSet.UpdatedStock, tableStocks)
		if err != nil {
			return wrapperr.WithStack(err)
		}
	}

	i3, err := txitem(changeSet.Outgoing, tableOutgoingEvents)
	if err != nil {
		return wrapperr.WithStack(err)
	}

	var txin dynamodb.TransactWriteItemsInput
	txin.TransactItems = []types.TransactWriteItem{*i1, *i2, *i3}

	txout, err := client.TransactWriteItems(ctx, &txin)
	if err != nil {
		return wrapperr.WithStack(err)
	}

	Logger.Infow(`transaction dynamodb result`, `result`, txout)

	return nil
}

func txitem(v interface{}, tableName string) (result *types.TransactWriteItem, err error) {
	var (
		mv map[string]types.AttributeValue
	)

	v = tomap(v)

	mv, err = attributevalue.MarshalMap(v)
	if err != nil {
		return nil, wrapperr.WithStack(err)
	}

	var put types.Put
	put.Item = mv
	put.TableName = &tableName

	result = &types.TransactWriteItem{}
	result.Put = &put

	return
}

func txitem2(v interface{}, tableName string, expr expression.Expression) (result *types.TransactWriteItem, err error) {
	var (
		mv map[string]types.AttributeValue
	)

	v = tomap(v)

	mv, err = attributevalue.MarshalMap(v)
	if err != nil {
		return nil, wrapperr.WithStack(err)
	}

	put := types.Put{
		ExpressionAttributeNames:  expr.Names(),
		ExpressionAttributeValues: expr.Values(),
		ConditionExpression:       expr.Condition(),
	}
	put.Item = mv
	put.TableName = &tableName

	result = &types.TransactWriteItem{}
	result.Put = &put

	return
}

func tomap(v interface{}) (result map[string]interface{}) {
	js, err := json.Marshal(v)
	if err != nil {
		panic(err)
	}

	result = make(map[string]interface{})
	err = json.Unmarshal(js, &result)
	if err != nil {
		panic(err)
	}

	return
}

var (
	tableIncomingEvents = os.Getenv("OPS_INCOMING_EVENTS")
	tableStocks         = os.Getenv("OPS_ENTITIES_TABLE")
	tableOutgoingEvents = os.Getenv("OPS_OUTGOING_EVENTS")
)

var (
	Logger *zap.SugaredLogger
)
